[BUG?] Can't use primitives with deserialization

### Expected Behavior

Logic for primitives should be the same as in jackson: fill primitives with their default values, if in json these fields are lost or nulls.

### Actual Behaviour

If you try to deserialize json without property, wich must me deserialized to primitive, you will see nex exception:
```
io.micronaut.serde.exceptions.SerdeException: Unable to deserialize type [ConfigItem E]: Null argument specified for [prop1]. If this argument is allowed to be null annotate it with @Nullable
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserialize(SpecificObjectDeserializer.java:279)
	at io.micronaut.serde.support.deserializers.CoreDeserializers$SpecificOnlyCollectionDeserializer$1.deserialize(CoreDeserializers.java:411)
	at io.micronaut.serde.support.deserializers.CoreDeserializers$SpecificOnlyCollectionDeserializer$1.deserialize(CoreDeserializers.java:400)
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserializeValue(SpecificObjectDeserializer.java:400)
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserialize(SpecificObjectDeserializer.java:174)
	at io.micronaut.serde.jackson.JacksonJsonMapper.readValue0(JacksonJsonMapper.java:131)
	at io.micronaut.serde.jackson.JacksonJsonMapper.readValue(JacksonJsonMapper.java:116)
	at io.micronaut.serde.jackson.JacksonJsonMapper.readValue(JacksonJsonMapper.java:160)
	at io.micronaut.serde.ObjectMapper.readValue(ObjectMapper.java:46)
	at com.micronaut.bug.config.MyEntityConfig.myConfigs(MyEntityConfig.java:36)
	at com.micronaut.bug.config.$MyEntityConfig$MyConfigs0$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
	at io.micronaut.context.DefaultBeanContext.getScopedBeanForDefinition(DefaultBeanContext.java:2922)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2822)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2782)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1638)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2035)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1281)
	at com.micronaut.bug.service.$MyEntityService$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3281)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3267)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2820)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2782)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1638)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2035)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1281)
	at com.micronaut.bug.controller.$MyEntityController$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3281)
	at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2664)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1932)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:237)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3453)
	at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3883)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:329)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:183)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:72)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:320)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:306)
	at com.micronaut.bug.Application.main(Application.java:23)
Caused by: io.micronaut.core.reflect.exception.InstantiationException: Null argument specified for [prop1]. If this argument is allowed to be null annotate it with @Nullable
	at io.micronaut.inject.beans.AbstractInitializableBeanIntrospection.instantiate(AbstractInitializableBeanIntrospection.java:226)
	at io.micronaut.core.beans.BeanIntrospection.instantiate(BeanIntrospection.java:81)
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserialize(SpecificObjectDeserializer.java:277)
	... 40 common frames omitted
``` 

### Steps To Reproduce

The same scenario and example like this https://github.com/micronaut-projects/micronaut-serialization/issues/168, but need to fix in files var1.json and var2.json field variant to string values.

ConfigData.java:
```java
@Serdeable
public record ConfigData(
        VariantEnum variant,
        List<String> params,
        List<Double> params2,
        List<ConfigItem> items
) {

    @Serdeable
    public record ConfigItem(
            boolean prop1,
            double prop2
    ) {
    }
}
```
var1.json:
```json
{
  "variant": "VAR1,
  "params": ["val1", "val2"],
  "params2": [122.23, 2222.56],
  "items": [
    {
      "prop1": true,
      "prop2": 3434.34
    },
    {
      "prop2": 3434.34
    }
  ]
}
```

Problem in second item oject, because there is no property `prop1` we will see exception:
```
io.micronaut.serde.exceptions.SerdeException: Unable to deserialize type [ConfigItem E]: Null argument specified for [prop1]. If this argument is allowed to be null annotate it with @Nullable
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserialize(SpecificObjectDeserializer.java:279)
	at io.micronaut.serde.support.deserializers.CoreDeserializers$SpecificOnlyCollectionDeserializer$1.deserialize(CoreDeserializers.java:411)
	at io.micronaut.serde.support.deserializers.CoreDeserializers$SpecificOnlyCollectionDeserializer$1.deserialize(CoreDeserializers.java:400)
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserializeValue(SpecificObjectDeserializer.java:400)
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserialize(SpecificObjectDeserializer.java:174)
	at io.micronaut.serde.jackson.JacksonJsonMapper.readValue0(JacksonJsonMapper.java:131)
	at io.micronaut.serde.jackson.JacksonJsonMapper.readValue(JacksonJsonMapper.java:116)
	at io.micronaut.serde.jackson.JacksonJsonMapper.readValue(JacksonJsonMapper.java:160)
	at io.micronaut.serde.ObjectMapper.readValue(ObjectMapper.java:46)
	at com.micronaut.bug.config.MyEntityConfig.myConfigs(MyEntityConfig.java:36)
	at com.micronaut.bug.config.$MyEntityConfig$MyConfigs0$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
	at io.micronaut.context.DefaultBeanContext.getScopedBeanForDefinition(DefaultBeanContext.java:2922)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2822)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2782)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1638)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2035)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1281)
	at com.micronaut.bug.service.$MyEntityService$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3281)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3267)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2820)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2782)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1638)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2035)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1281)
	at com.micronaut.bug.controller.$MyEntityController$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2336)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3281)
	at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2664)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1932)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:237)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3453)
	at io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3883)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:329)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:183)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:72)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:320)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:306)
	at com.micronaut.bug.Application.main(Application.java:23)
Caused by: io.micronaut.core.reflect.exception.InstantiationException: Null argument specified for [prop1]. If this argument is allowed to be null annotate it with @Nullable
	at io.micronaut.inject.beans.AbstractInitializableBeanIntrospection.instantiate(AbstractInitializableBeanIntrospection.java:226)
	at io.micronaut.core.beans.BeanIntrospection.instantiate(BeanIntrospection.java:81)
	at io.micronaut.serde.support.deserializers.SpecificObjectDeserializer.deserialize(SpecificObjectDeserializer.java:277)
	... 40 common frames omitted
``` 
I think the more correct logic for primitives should be the same as in jackson: fill primitives with their default values.

### Environment Information

 - Windows 11
 - JDK17

### Example Application

https://github.com/altro3/micronaut3-bug

### Version

3.4.0

--- a/serde-api/src/main/java/io/micronaut/serde/config/annotation/SerdeConfig.java
+++ b/serde-api/src/main/java/io/micronaut/serde/config/annotation/SerdeConfig.java
@@ -68,6 +68,11 @@ public @interface SerdeConfig {
     String PROPERTY = "property";
 
     /**
+     * Whether this property is required (must be present in the input).
+     */
+    String REQUIRED = "required";
+
+    /**
      * Is it ignored.
      */
     String IGNORED = "ignored";

--- a/serde-jackson/src/test/groovy/io/micronaut/serde/jackson/annotation/JsonPropertySpec.groovy
+++ b/serde-jackson/src/test/groovy/io/micronaut/serde/jackson/annotation/JsonPropertySpec.groovy
@@ -204,13 +204,13 @@ import io.micronaut.serde.annotation.Serdeable;
 
 @Serdeable
 class Test {
-    private final int value;
+    private final $type value;
     
-    Test(int value) {
+    Test($type value) {
         this.value = value;
     }
 
-    public int getValue() {
+    public $type getValue() {
         return value;
     }
 }
@@ -224,6 +224,15 @@ class Test {
 
         cleanup:
         ctx.close()
+
+        where:
+        type      | value
+        "byte" | (byte) 0
+        "short" | (short) 0
+        "int" | 0
+        "long" | 0L
+        "float" | 0F
+        "double" | 0D
     }
 
     @Unroll

--- a/serde-jackson/src/test/groovy/io/micronaut/serde/jackson/annotation/JsonPropertySpec.groovy
+++ b/serde-jackson/src/test/groovy/io/micronaut/serde/jackson/annotation/JsonPropertySpec.groovy
@@ -1,6 +1,7 @@
 package io.micronaut.serde.jackson.annotation
 
 import io.micronaut.core.beans.exceptions.IntrospectionException
+import io.micronaut.serde.exceptions.SerdeException
 import io.micronaut.serde.jackson.JsonCompileSpec
 import spock.lang.Requires
 import spock.lang.Unroll
@@ -123,6 +124,106 @@ class Test {
         type    | value
         Integer | 'junk'
         URL     | 'ws://junk'
+    }
+
+    void "test required primitive field"() {
+
+        given:
+        def ctx = buildContext('test.Test', """
+package test;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.micronaut.serde.annotation.Serdeable;
+
+@Serdeable
+class Test {
+    @JsonProperty(required = true)
+    private int value;
+
+    Test(int value) {
+        this.value = value;
+    }
+
+    public int getValue() {
+        return value;
+    }
+}
+""")
+
+        when:
+        def bean =
+                jsonMapper.readValue('{}', argumentOf(ctx, 'test.Test'))
+        then:
+        def e = thrown(SerdeException)
+        e.message.contains("Unable to deserialize type [test.Test]. Required constructor parameter [int value] at index [0] is not present or is null in the supplied data")
+
+        cleanup:
+        ctx.close()
+    }
+
+    void "test optional by default primitive field"() {
+
+        given:
+        def ctx = buildContext('test.Test', """
+package test;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.micronaut.serde.annotation.Serdeable;
+
+@Serdeable
+class Test {
+    private int value = 5;
+
+    public void setValue(int value) {
+        this.value = value;
+    } 
+    public int getValue() {
+        return value;
+    }
+}
+""")
+
+        when:
+        def bean =
+                jsonMapper.readValue('{}', argumentOf(ctx, 'test.Test'))
+        then:
+        bean.value == 5
+
+        cleanup:
+        ctx.close()
+    }
+
+    void "test optional by default primitive field in constructor"() {
+
+        given:
+        def ctx = buildContext('test.Test', """
+package test;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.micronaut.serde.annotation.Serdeable;
+
+@Serdeable
+class Test {
+    private final int value;
+    
+    Test(int value) {
+        this.value = value;
+    }
+
+    public int getValue() {
+        return value;
+    }
+}
+""")
+
+        when:
+        def bean =
+                jsonMapper.readValue('{}', argumentOf(ctx, 'test.Test'))
+        then:
+        bean.value == 0
+
+        cleanup:
+        ctx.close()
     }
 
     @Unroll

--- a/serde-jackson/src/test/groovy/io/micronaut/serde/jackson/annotation/JsonPropertySpec.groovy
+++ b/serde-jackson/src/test/groovy/io/micronaut/serde/jackson/annotation/JsonPropertySpec.groovy
@@ -201,6 +201,7 @@ package test;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 import io.micronaut.serde.annotation.Serdeable;
+import io.micronaut.core.annotation.Nullable;
 
 @Serdeable
 class Test {
@@ -220,19 +221,26 @@ class Test {
         def bean =
                 jsonMapper.readValue('{}', argumentOf(ctx, 'test.Test'))
         then:
-        bean.value == 0
+        bean.value == value
 
         cleanup:
         ctx.close()
 
         where:
         type      | value
-        "byte" | (byte) 0
-        "short" | (short) 0
-        "int" | 0
-        "long" | 0L
-        "float" | 0F
-        "double" | 0D
+        "byte"    | (byte) 0
+        "short"   | (short) 0
+        "int"     | 0
+        "long"    | 0L
+        "float"   | 0F
+        "double"  | 0D
+
+        "@Nullable Byte"    | null
+        "@Nullable Short"   | null
+        "@Nullable Integer" | null
+        "@Nullable Long"    | null
+        "@Nullable Float"   | null
+        "@Nullable Double"  | null
     }
 
     @Unroll

--- a/serde-processor/src/main/java/io/micronaut/serde/processor/jackson/JsonPropertyMapper.java
+++ b/serde-processor/src/main/java/io/micronaut/serde/processor/jackson/JsonPropertyMapper.java
@@ -60,6 +60,9 @@ public class JsonPropertyMapper implements NamedAnnotationMapper {
                 // no-op
             }
         }
+        if (annotation.booleanValue("required").orElse(false)) {
+            builder.member(SerdeConfig.REQUIRED, true);
+        }
         values.add(builder.build());
         return values;
     }

--- a/serde-support/src/main/java/io/micronaut/serde/support/DefaultSerdeRegistry.java
+++ b/serde-support/src/main/java/io/micronaut/serde/support/DefaultSerdeRegistry.java
@@ -20,6 +20,7 @@ import io.micronaut.context.BeanRegistration;
 import io.micronaut.context.annotation.Secondary;
 import io.micronaut.context.exceptions.ConfigurationException;
 import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
 import io.micronaut.core.annotation.Order;
 import io.micronaut.core.beans.BeanIntrospection;
 import io.micronaut.core.order.OrderUtil;
@@ -447,6 +448,12 @@ public class DefaultSerdeRegistry implements SerdeRegistry {
                     getType(), Argument.BYTE
             );
         }
+
+        @Nullable
+        @Override
+        public Byte getDefaultValue(@NonNull DecoderContext context, @NonNull Argument<? super Byte> type) {
+            return type.isPrimitive() ? (byte) 0 : null;
+        }
     }
 
     private static final class DoubleSerde extends SerdeRegistrar<Double> implements NumberSerde<Double> {
@@ -474,6 +481,12 @@ public class DefaultSerdeRegistry implements SerdeRegistry {
             return Arrays.asList(
                     getType(), Argument.DOUBLE
             );
+        }
+
+        @Nullable
+        @Override
+        public Double getDefaultValue(@NonNull DecoderContext context, @NonNull Argument<? super Double> type) {
+            return type.isPrimitive() ? 0D : null;
         }
     }
 
@@ -503,6 +516,12 @@ public class DefaultSerdeRegistry implements SerdeRegistry {
                     getType(), Argument.SHORT
             );
         }
+
+        @Nullable
+        @Override
+        public Short getDefaultValue(@NonNull DecoderContext context, @NonNull Argument<? super Short> type) {
+            return type.isPrimitive() ? (short) 0 : null;
+        }
     }
 
     private static final class FloatSerde extends SerdeRegistrar<Float> implements NumberSerde<Float> {
@@ -530,6 +549,12 @@ public class DefaultSerdeRegistry implements SerdeRegistry {
             return Arrays.asList(
                     getType(), Argument.FLOAT
             );
+        }
+
+        @Nullable
+        @Override
+        public Float getDefaultValue(@NonNull DecoderContext context, @NonNull Argument<? super Float> type) {
+            return type.isPrimitive() ? 0F : null;
         }
     }
 
@@ -559,6 +584,12 @@ public class DefaultSerdeRegistry implements SerdeRegistry {
                     getType(), Argument.INT
             );
         }
+
+        @Nullable
+        @Override
+        public Integer getDefaultValue(@NonNull DecoderContext context, @NonNull Argument<? super Integer> type) {
+            return type.isPrimitive() ? 0 : null;
+        }
     }
 
     private static final class LongSerde extends SerdeRegistrar<Long> implements NumberSerde<Long> {
@@ -586,6 +617,12 @@ public class DefaultSerdeRegistry implements SerdeRegistry {
             return Arrays.asList(
                     getType(), Argument.LONG
             );
+        }
+
+        @Nullable
+        @Override
+        public Long getDefaultValue(@NonNull DecoderContext context, @NonNull Argument<? super Long> type) {
+            return type.isPrimitive() ? 0L : null;
         }
     }
 

--- a/serde-support/src/main/java/io/micronaut/serde/support/deserializers/DeserBean.java
+++ b/serde-support/src/main/java/io/micronaut/serde/support/deserializers/DeserBean.java
@@ -664,31 +664,20 @@ class DeserBean<T> {
 
         public void setDefault(Deserializer.DecoderContext decoderContext, @NonNull B bean) throws SerdeException {
             if (!explicitlyRequired) {
-                if (defaultValue != null) {
+                P def = defaultValue;
+                if (def == null) {
+                    if (!mustSetField) {
+                        return;
+                    }
+                    def = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
+                }
+                if (def != null) {
                     if (beanProperty != null) {
-                        beanProperty.setUnsafe(bean, defaultValue);
+                        beanProperty.setUnsafe(bean, def);
                         return;
                     }
                     if (beanMethod != null) {
-                        beanMethod.invoke(bean, defaultValue);
-                        return;
-                    }
-                }
-
-                if (!mustSetField) {
-                    return;
-                }
-                if (beanProperty != null) {
-                    P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
-                    if (newDefaultValue != null) {
-                        beanProperty.setUnsafe(bean, newDefaultValue);
-                        return;
-                    }
-                }
-                if (beanMethod != null) {
-                    P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
-                    if (newDefaultValue != null) {
-                        beanMethod.invoke(bean, newDefaultValue);
+                        beanMethod.invoke(bean, def);
                         return;
                     }
                 }

--- a/serde-support/src/main/java/io/micronaut/serde/support/deserializers/DeserBean.java
+++ b/serde-support/src/main/java/io/micronaut/serde/support/deserializers/DeserBean.java
@@ -574,7 +574,8 @@ class DeserBean<T> {
         public final Argument<P> argument;
         @Nullable
         public final P defaultValue;
-        public final boolean required;
+        public final boolean mustSetField;
+        public final boolean explicitlyRequired;
         public final boolean nonNull;
         public final boolean nullable;
         public final boolean isAnySetter;
@@ -625,7 +626,7 @@ class DeserBean<T> {
             this.instrospection = instrospection;
             this.index = index;
             this.argument = argument;
-            this.required = argument.isNonNull() || argument.isAssignableFrom(Optional.class)
+            this.mustSetField = argument.isNonNull() || argument.isAssignableFrom(Optional.class)
                     || argument.isAssignableFrom(OptionalLong.class)
                     || argument.isAssignableFrom(OptionalDouble.class)
                     || argument.isAssignableFrom(OptionalInt.class);
@@ -657,35 +658,39 @@ class DeserBean<T> {
                     .orElse(null);
             this.backRef = annotationMetadata.stringValue(SerdeConfig.SerBackRef.class)
                     .orElse(null);
+            this.explicitlyRequired = annotationMetadata.booleanValue(SerdeConfig.class, SerdeConfig.REQUIRED)
+                    .orElse(false);
         }
 
         public void setDefault(Deserializer.DecoderContext decoderContext, @NonNull B bean) throws SerdeException {
-            if (defaultValue != null) {
-                if (beanProperty != null) {
-                    beanProperty.setUnsafe(bean, defaultValue);
+            if (!explicitlyRequired) {
+                if (defaultValue != null) {
+                    if (beanProperty != null) {
+                        beanProperty.setUnsafe(bean, defaultValue);
+                        return;
+                    }
+                    if (beanMethod != null) {
+                        beanMethod.invoke(bean, defaultValue);
+                        return;
+                    }
+                }
+
+                if (!mustSetField) {
                     return;
+                }
+                if (beanProperty != null) {
+                    P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
+                    if (newDefaultValue != null) {
+                        beanProperty.setUnsafe(bean, newDefaultValue);
+                        return;
+                    }
                 }
                 if (beanMethod != null) {
-                    beanMethod.invoke(bean, defaultValue);
-                    return;
-                }
-            }
-
-            if (!required) {
-                return;
-            }
-            if (beanProperty != null) {
-                P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
-                if (newDefaultValue != null) {
-                    beanProperty.setUnsafe(bean, newDefaultValue);
-                    return;
-                }
-            }
-            if (beanMethod != null) {
-                P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
-                if (newDefaultValue != null) {
-                    beanMethod.invoke(bean, newDefaultValue);
-                    return;
+                    P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
+                    if (newDefaultValue != null) {
+                        beanMethod.invoke(bean, newDefaultValue);
+                        return;
+                    }
                 }
             }
             throw new SerdeException("Unable to deserialize type [" + instrospection.getBeanType().getName() + "]. Required property [" + argument +
@@ -693,17 +698,19 @@ class DeserBean<T> {
         }
 
         public void setDefault(Deserializer.DecoderContext decoderContext, @NonNull Object[] params) throws SerdeException {
-            if (defaultValue != null) {
-                params[index] = defaultValue;
-                return;
-            }
-            if (!required) {
-                return;
-            }
-            P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
-            if (newDefaultValue != null) {
-                params[index] = newDefaultValue;
-                return;
+            if (!explicitlyRequired) {
+                if (defaultValue != null) {
+                    params[index] = defaultValue;
+                    return;
+                }
+                if (!mustSetField && !argument.isPrimitive()) {
+                    return;
+                }
+                P newDefaultValue = (P) deserializer.getDefaultValue(decoderContext, (Argument) argument);
+                if (newDefaultValue != null) {
+                    params[index] = newDefaultValue;
+                    return;
+                }
             }
             throw new SerdeException("Unable to deserialize type [" + instrospection.getBeanType().getName() + "]. Required constructor parameter [" + argument + "] at index [" + index + "] is not present or is null in the supplied data");
         }

--- a/serde-support/src/main/java/io/micronaut/serde/support/deserializers/SpecificObjectDeserializer.java
+++ b/serde-support/src/main/java/io/micronaut/serde/support/deserializers/SpecificObjectDeserializer.java
@@ -655,7 +655,7 @@ class SpecificObjectDeserializer implements Deserializer<Object>, UpdatingDeseri
                     if (!satisfied) {
                         if (der.defaultValue != null) {
                             params[der.index] = der.defaultValue;
-                        } else if (der.required) {
+                        } else if (der.mustSetField) {
                             throw new SerdeException("Unable to deserialize type [" + unwrapped.introspection.getBeanType() + "]. Required constructor parameter [" + der.argument + "] at index [" + der.index + "] is not present in supplied data");
 
                         }

